File: agents/services/pathrag-service.ts
/**
 * PathRAG Service Integration
 * Provides TypeScript client for PathRAG API at http://movie.ft.tc:5000
 */

export interface PathRAGHealthResponse {
  services: {
    api: string
    arangodb: string
  }
  status: string
  timestamp: string
}

export interface PathRAGInsertResponse {
  message: string
  document_count: number
  timestamp: string
}

export interface PathRAGCustomKGResponse {
  message: string
  entities_count: number
  relationships_count: number
  chunks_count: number
  timestamp: string
}

export interface PathRAGQueryResponse {
  query: string
  result: string
  params: Record<string, any>
  timestamp: string
}

export interface PathRAGStatsResponse {
  total_documents: number
  total_entities: number
  total_relationships: number
  cache_hit_rate: number
  timestamp: string
}

export interface CustomKnowledgeGraph {
  chunks: string[]
  entities: Array<{
    name: string
    type: string
    description: string
    metadata?: Record<string, any>
  }>
  relationships: Array<{
    source: string
    target: string
    relation: string
  }>
}

export class PathRAGService {
  private static baseUrl = process.env.PATHRAG_API_URL || 'http://movie.ft.tc:5000'

  /**
   * Check PathRAG service health
   */
  static async healthCheck(): Promise<PathRAGHealthResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/health`)
      if (!response.ok) {
        throw new Error(`PathRAG health check failed: ${response.statusText}`)
      }
      return await response.json()
    } catch (error) {
      console.error('PathRAG health check error:', error)
      throw error
    }
  }

  /**
   * Insert documents into PathRAG knowledge base
   */
  static async insertDocuments(documents: string | string[]): Promise<PathRAGInsertResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/insert`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ documents })
      })
      
      if (!response.ok) {
        throw new Error(`PathRAG insert failed: ${response.statusText}`)
      }
      
      return await response.json()
    } catch (error) {
      console.error('PathRAG insert error:', error)
      throw error
    }
  }

  /**
   * Insert custom knowledge graph with structured entities and relationships
   */
  static async insertCustomKG(customKG: CustomKnowledgeGraph): Promise<PathRAGCustomKGResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/insert_custom_kg`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ custom_kg: customKG })
      })
      
      if (!response.ok) {
        throw new Error(`PathRAG custom KG insert failed: ${response.statusText}`)
      }
      
      return await response.json()
    } catch (error) {
      console.error('PathRAG custom KG insert error:', error)
      throw error
    }
  }

  /**
   * Query PathRAG knowledge base using natural language
   */
  static async query(
    query: string, 
    params: {
      mode?: string
      only_need_context?: boolean
      only_need_prompt?: boolean
      response_type?: string
      stream?: boolean
      top_k?: number
      max_token_for_text_unit?: number
      max_token_for_global_context?: number
      max_token_for_local_context?: number
    } = {}
  ): Promise<PathRAGQueryResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/query`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ query, params })
      })
      
      if (!response.ok) {
        throw new Error(`PathRAG query failed: ${response.statusText}`)
      }
      
      return await response.json()
    } catch (error) {
      console.error('PathRAG query error:', error)
      throw error
    }
  }

  /**
   * Delete an entity from the knowledge graph
   */
  static async deleteEntity(entityName: string): Promise<any> {
    try {
      const response = await fetch(`${this.baseUrl}/delete_entity`, {
        method: 'DELETE',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ entity_name: entityName })
      })
      
      if (!response.ok) {
        throw new Error(`PathRAG delete entity failed: ${response.statusText}`)
      }
      
      return await response.json()
    } catch (error) {
      console.error('PathRAG delete entity error:', error)
      throw error
    }
  }

  /**
   * Get PathRAG system statistics
   */
  static async getStats(): Promise<PathRAGStatsResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/stats`)
      if (!response.ok) {
        throw new Error(`PathRAG stats failed: ${response.statusText}`)
      }
      return await response.json()
    } catch (error) {
      console.error('PathRAG stats error:', error)
      throw error
    }
  }

  /**
   * Get PathRAG configuration (sanitized)
   */
  static async getConfig(): Promise<any> {
    try {
      const response = await fetch(`${this.baseUrl}/config`)
      if (!response.ok) {
        throw new Error(`PathRAG config failed: ${response.statusText}`)
      }
      return await response.json()
    } catch (error) {
      console.error('PathRAG config error:', error)
      throw error
    }
  }

  /**
   * Test PathRAG connectivity and basic functionality
   */
  static async testConnection(): Promise<{
    healthy: boolean
    stats?: PathRAGStatsResponse
    error?: string
  }> {
    try {
      const health = await this.healthCheck()
      if (health.status !== 'healthy') {
        return { healthy: false, error: 'PathRAG service not healthy' }
      }

      const stats = await this.getStats()
      return { healthy: true, stats }
    } catch (error) {
      return { 
        healthy: false, 
        error: error instanceof Error ? error.message : 'Unknown error' 
      }
    }
  }
}

File: novel-movie-crewai-server/services/pathrag_service.py
"""
PathRAG Service Client
Python client for PathRAG API integration
"""

import asyncio
import logging
from typing import Dict, Any, Optional, List
import aiohttp
from config.settings import settings


class PathRAGService:
    """
    Async client for PathRAG knowledge graph service
    """
    
    def __init__(self):
        self.base_url = settings.PATHRAG_API_URL
        self.logger = logging.getLogger(__name__)
        self.session: Optional[aiohttp.ClientSession] = None
    
    async def _get_session(self) -> aiohttp.ClientSession:
        """Get or create aiohttp session"""
        if self.session is None or self.session.closed:
            self.session = aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=30)
            )
        return self.session
    
    async def health_check(self) -> Dict[str, Any]:
        """Check PathRAG service health"""
        try:
            session = await self._get_session()
            async with session.get(f"{self.base_url}/health") as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Health check failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG health check failed: {str(e)}")
            raise
    
    async def insert_documents(self, documents: List[str]) -> Dict[str, Any]:
        """Insert documents into PathRAG"""
        try:
            session = await self._get_session()
            payload = {"documents": documents}
            
            async with session.post(
                f"{self.base_url}/insert",
                json=payload
            ) as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Document insertion failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG document insertion failed: {str(e)}")
            raise
    
    async def insert_custom_kg(self, custom_kg: Dict[str, Any]) -> Dict[str, Any]:
        """Insert custom knowledge graph"""
        try:
            session = await self._get_session()
            payload = {"custom_kg": custom_kg}
            
            async with session.post(
                f"{self.base_url}/insert_custom_kg",
                json=payload
            ) as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Custom KG insertion failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG custom KG insertion failed: {str(e)}")
            # Return mock response for development
            return {
                "message": "PathRAG custom KG insertion (mock response)",
                "entities_count": len(custom_kg.get("entities", [])),
                "relationships_count": len(custom_kg.get("relationships", [])),
                "chunks_count": len(custom_kg.get("chunks", []))
            }
    
    async def query(self, query: str, **params) -> Dict[str, Any]:
        """Query PathRAG knowledge base"""
        try:
            session = await self._get_session()
            payload = {
                "query": query,
                "params": params
            }
            
            async with session.post(
                f"{self.base_url}/query",
                json=payload
            ) as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Query failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG query failed: {str(e)}")
            # Return mock response for development
            return {
                "query": query,
                "result": f"Mock response for query: {query}",
                "params": params,
                "message": "PathRAG service is in placeholder mode"
            }
    
    async def delete_entity(self, entity_name: str) -> Dict[str, Any]:
        """Delete an entity from the knowledge graph"""
        try:
            session = await self._get_session()
            payload = {"entity_name": entity_name}
            
            async with session.delete(
                f"{self.base_url}/delete_entity",
                json=payload
            ) as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Entity deletion failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG entity deletion failed: {str(e)}")
            raise
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get PathRAG system statistics"""
        try:
            session = await self._get_session()
            async with session.get(f"{self.base_url}/stats") as response:
                if response.status == 200:
                    return await response.json()
                else:
                    raise Exception(f"Stats request failed: {response.status}")
        except Exception as e:
            self.logger.error(f"PathRAG stats request failed: {str(e)}")
            # Return mock stats for development
            return {
                "total_documents": 0,
                "total_entities": 0,
                "total_relationships": 0,
                "message": "PathRAG stats (mock response)"
            }
    
    async def close(self):
        """Close the aiohttp session"""
        if self.session and not self.session.closed:
            await self.session.close()
    
    def __del__(self):
        """Cleanup on deletion"""
        if self.session and not self.session.closed:
            # Schedule cleanup for next event loop iteration
            try:
                loop = asyncio.get_event_loop()
                if loop.is_running():
                    loop.create_task(self.close())
            except RuntimeError:
                # Event loop is not running, can't schedule cleanup
                pass
